utils/http/proxy/proxy.go

```diff
@@ -0,0 +1,153 @@
+package proxy
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/go-http-utils/headers"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	httpheaders "github.com/ARM-software/golang-utils/utils/http/headers"
+	"github.com/ARM-software/golang-utils/utils/reflection"
+	"github.com/ARM-software/golang-utils/utils/safecast"
+	"github.com/ARM-software/golang-utils/utils/safeio"
+)
+
+// ProxyDisallowList describes headers which are not proxied back.
+var ProxyDisallowList = []string{
+	headers.AccessControlAllowOrigin,
+	headers.AccessControlAllowMethods,
+	headers.AccessControlAllowHeaders,
+	headers.AccessControlExposeHeaders,
+	headers.AccessControlMaxAge,
+	headers.AccessControlAllowCredentials,
+}
+
+// ProxyRequest proxies a request to a new endpoint. The method can also be changed. Headers are sanitised during the process.
+func ProxyRequest(r *http.Request, proxyMethod, endpoint string) (proxiedRequest *http.Request, err error) {
+	if reflection.IsEmpty(r) {
+		err = commonerrors.UndefinedVariable("request to proxy")
+		return
+	}
+	ctx := r.Context()
+	contentLength := determineRequestContentLength(r)
+	h := httpheaders.FromRequest(r).AllowList(headers.Authorization)
+	if reflection.IsEmpty(proxyMethod) {
+		proxyMethod = http.MethodGet
+	}
+	proxiedRequest, err = http.NewRequestWithContext(ctx, proxyMethod, endpoint, r.Body)
```

Comment:
If the origin returns a redirect, say 301, will this natively follow the 301 to resolution or just return the 301 back to the client to follow?

Reply:
this does not make the call this is just creating the request object (the configuration object you were suggesting)
It is rewriting the endpoint path and method but it also needs to "clone" quite a few things manually such as headers and body